Fix #3417: Make region names editable in-place

juansaba opened the pull request:
Aside from implementing the feature described in the issue, I also did the following things:

- Removed the temporary list of labels {labelsList} since it was not necessary.
- Updated the resetEditor() function so that it would reset ALL the variables in the $scope. Previously, only a few variables were cleared but many other variables remained with their old values, so when the user when back and select a different image, the leftovers from the old image and region lists would cause all sorts of trouble.
- Fixed the "duplicate label" message. There was a bug where this message would persist even after fixing the problem. This also happened after resetEditor() and then selecting a new image.
- I also updated the image displayed when selecting type of interaction, that used to show two regions called "1" and "2". I replaced it with regions "Earth" and "Saturn", which make more sense for the actual image being displayed and sends the message to the user that regions should have meaningful names (see final image below).
- Finally, this PR also makes the "Save Interaction" button disabled if there are customization args warnings.

![image](https://cloud.githubusercontent.com/assets/4760286/26018306/d908ca16-3722-11e7-9abe-dcb64ea4f77d.png)

juansaba committed:
Allow editing region names in-place

juansaba committed:
Merge remote-tracking branch 'upstream/develop' into images-region-names

juansaba committed:
Making save interaction button disabled if there are customization args warnings

juansaba committed:
Fixed linter problems

seanlip commented:
Thanks, @juansaba! Just to check, in addition to the bullet points you listed, this PR also makes the customization args unsaveable if they are invalid, right (for all interactions)?

seanlip commented:
(I ask because we may need to think about displaying an error message if the "save" button is disabled, so that the user knows what they need to do to enable it. Otherwise it could be frustrating. E.g. if you go to multiple choice, and accidentally type the same text for two different choices, and the save button is grayed out but there's no indication of what you did wrong.)

juansaba commented:
Right.Right now you WILL see an error message, saying that there are duplicate region labels, but the showing of the error and the disabling of the button are done by different (independent) logic.Perhaps it would be best to add a generic error message tied directly to the logic of the save enable/disable of the button so that every time you see the button disabled you'll also see the error (the list of warnings, basically).Is that what you were thinking?I can certainly do something like that, which would work for all interactions.On May 20, 2017 2:06 PM, Sean Lip <notifications@github.com> wrote:(I ask because we may need to think about displaying an error message if the "save" button is disabled, so that the user knows what they need to do to enable it.)

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

seanlip commented:
Wow, looks like your email reply got mangled a bit in the [thread](https://github.com/oppia/oppia/pull/3449#issuecomment-302905346)! First time I've seen that...

Yeah, I was thinking something generic. Maybe the text of the first warning? Don't know how it should work though -- it could be a tooltip/popover on the button, or it could just be in plain sight -- your call.

juansaba commented:
First time I replied on email to post on the PR too.Yeah I was thinking tooltip overlay on the button showing ALL of the warnings, since all of them would prevent the saving. If you just show the first one you get that annoying experience of fixing one thing to find out about the next.Do you have an example of an overlay done in Oppia? Otherwise I'll figure it out.I'll send another commit for this ;)On May 20, 2017 5:36 PM, Sean Lip <notifications@github.com> wrote:Wow, looks like your email reply got mangled a bit in the thread! First time I've seen that...
Yeah, I was thinking something generic. Maybe the text of the first warning? Don't know how it should work though -- it could be a tooltip/popover on the button, or it could just be in plain sight -- your call.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

seanlip commented:
Hm, yeah, that's true (re multiple messages). OK, your call!

I don't think we have a standard pattern for this sort of thing yet, so I would suggest going with whatever looks good and is clearly readable.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -44,13 +44,13 @@ oppia.factory('interactionDetailsCache', [function() {
+  'stateCustomizationArgsService', 'editabilityService',
+  'explorationStatesService', 'graphDataService', 'interactionDetailsCache',
+  'oppiaExplorationHtmlFormatterService', 'UrlInterpolationService',
+  function($scope, $http, $rootScope, $modal, $injector, $filter, alertsService,
```
Nit: see if you can make the line breaks here match up with those above. It makes it easier to compare the two lists visually.

juansaba replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -186,12 +186,12 @@ oppia.controller('StateInteraction', [
+            'UrlInterpolationService', 'editorFirstTimeEventsService',
             function(
-                $scope, $modalInstance, stateInteractionIdService,
+                $scope, $modalInstance, $injector, stateInteractionIdService,
```
Ditto, re line breaks.

juansaba replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -244,6 +244,21 @@ oppia.controller('StateInteraction', [
+                }
+                var validationService =
+                  $injector.get(
+                    INTERACTION_SPECS[
```
Consider making a new variable for the name of the service, otherwise the indentation starts to look a bit crazy :)

juansaba replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -244,6 +244,21 @@ oppia.controller('StateInteraction', [
+                var warningsList =
+                  validationService.getCustomizationArgsWarnings(
+                    stateCustomizationArgsService.displayed);
+                return warningsList.length == 0;
```
===

juansaba replied:
@seanlip Ok, I assume it's a rule in Oppia to always use === and !== for everything? I don't think in this case there could be any type conversion problem, since we know the types of both sides of the condition. But if it's a rule, I'll follow it.

seanlip replied:
Yep! We generally use === and !== (otherwise I think you'll fail the frontend linter).

juansaba replied:
Ok, there may be a problem in the linter then, cause I always run the frontend tests and the linter is run when pushing, and I didn't get an error on this. Just saying ;)

seanlip commented on extensions/objects/templates/ImageWithRegionsEditor.js:
```diff
@@ -56,67 +56,80 @@ oppia.directive('imageWithRegionsEditor', [
 
           // Dynamically defines the CSS style for the region label.
           $scope.getRegionLabelStyle = function(index) {
+            var commonStyles = 'font-size: 14px; pointer-events: none;';
```
Nit: add a space at the end of the string.

juansaba replied:
Done (sort of).

seanlip replied:
Works for me :)

seanlip commented on extensions/objects/templates/ImageWithRegionsEditor.js:
```diff
@@ -56,67 +56,80 @@ oppia.directive('imageWithRegionsEditor', [
-              return region.label;
-            }
-          );
+          $scope.setInitialState = function() {
```
"State" is a term of art in Oppia, and means something different (a card). Let's call this $scope.initialize() or $scope.initEditor().

juansaba replied:
Done. (initializeEditor?)

seanlip replied:
Yep, sounds good.

seanlip commented on extensions/objects/templates/image_with_regions_editor.html:
```diff
@@ -80,3 +78,15 @@ <h4 class="modal-title">
   </div>
 </script>
+<style type='text/css'>
+  input.region-label-editor {
```
Use a more specific class that's scoped to this directive, so that there's no danger of it applying outside the scope of this file. Perhaps `image-with-regions-editor input.region-label-editor` might work? (But you'd need to double-check that.)

juansaba replied:
I copied from the class on the svg item, to add the "oppia-image-with-regions-" prefix, so the full name is now "oppia-image-with-regions-region-label-editor".

seanlip replied:
Hm, I guess that works too!

seanlip commented on extensions/objects/templates/image_with_regions_editor.html:
```diff
@@ -80,3 +78,15 @@ <h4 class="modal-title">
     <button type="button" ng-click="confirmClear()" class="btn btn-danger">Clear Image and Regions</button>
   </div>
 </script>
+<style type='text/css'>
```
Just `<style>` will do.

juansaba replied:
Done.

seanlip commented on extensions/objects/templates/image_with_regions_editor.html:
```diff
@@ -56,21 +56,19 @@
-    <h4 class="modal-title">
-      <span>Clear Image and Regions</span>
-    </h4>
+    <h4 class="modal-title"><span>Clear Image and Regions</span></h4>
```
Not sure if inner span is needed, FYI. (I know it was there before.)

juansaba replied:
Done.

seanlip reviewed (changes requested):
Thanks, @juansaba, looks good! Just a few comments (besides the general one I mentioned earlier).

juansaba committed:
Addressing latest review comments

juansaba commented:
@seanlip Ok, I've addressed the comments.

In terms of the Save button warnings, see below some screenshots of what it looks like with my implementation. Let me know what you think. Also, I tested the other interactions as well to make sure there was nothing too specific I assumed in the code, since this feature should work on all interactions.

**Screenshots below:**

As soon as you load the editor:

![image](https://cloud.githubusercontent.com/assets/4760286/26292705/fd7068ae-3e6c-11e7-9af3-a1426fbd6924.png)

Once you select the image:

![image](https://cloud.githubusercontent.com/assets/4760286/26292725/2d12366e-3e6d-11e7-8ff4-82e1df193864.png)

Then you create a couple regions and it goes back to normal:

![image](https://cloud.githubusercontent.com/assets/4760286/26292738/4847cd2c-3e6d-11e7-9641-17b7609a795e.png)

Then when you generate a duplicate, you'll see the following:

![image](https://cloud.githubusercontent.com/assets/4760286/26292757/70b564d6-3e6d-11e7-84d4-0e6123b237c2.png)

I think it makes sense to keep the red message in the editor as it is the first (direct) feedback the user gets and it's specific to the images and regions editor. I think in this case it is ok that it may be somewhat redundant with the alerts on the save button.

juansaba commented:
@seanlip Oh and you'll notice I went for standard tooltips (title="...") instead of a fancy overlay. I think in this case, less is more. Let me know what you think.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -287,6 +308,48 @@ oppia.controller('StateInteraction', [
                 stateCustomizationArgsService.displayed = {};
               };
 
+              $scope.getSaveInteractionButtonStyleTooltip = function() {
```
Shouldn't this be getSaveInteractionButtonTooltip()?

juansaba replied:
Right! Done.

seanlip commented on extensions/objects/templates/image_with_regions_editor.html:
```diff
@@ -80,3 +78,15 @@ <h4 class="modal-title">
 </script>
+<style>
+  input.oppia-image-with-regions-region-label-editor {
+    position: absolute;
```
Nits:
- alphabetize CSS attrs (background-color, then border, then color, etc.)
- use 0 (no units) instead of 0px

juansaba replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -287,6 +308,48 @@ oppia.controller('StateInteraction', [
+                    'padding-left: 32px; ' +
+                    'background-repeat: no-repeat; ' +
+                    'background-image: ' +
+                    'url("../assets/common/images/general/warning.png")';
```
This is a bit messy (too much CSS in the JS) and also the condition is really the same as the ng-disabled one. Could we do two things:

(a) Create a $scope.doCustomizationArgsWarningsExist() function in the JS that does this:

    return !stateInteractionIdService.displayed || form.schemaForm.$invalid || !areCustomizationArgsValid()

(b) Do this stuff inline in the HTML with ng-class: 

    ng-class="{'class-name': doCustomizationArgsWarningsExist()}"

and something similar for ng-disabled.

seanlip replied:
(In fact -- doesn't areCustomizationArgsValid() take care of some of this already? Maybe we can move everything into that.)

juansaba replied:
I played with this for a long time on Sunday and some of these functions were being called even when some of the pre-requisites weren't met, despite the ng-if of the elements involved, so I got a bit paranoid and started checking all these conditions everywhere.

Nevertheless, I cleaned this up a bit now. Let me know what you think.

seanlip replied:
That's odd ... thanks for letting me know. I don't think consolidating (as described here) should affect that, but if you have more details on a repro case then that would be interesting to investigate.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -287,6 +308,48 @@ oppia.controller('StateInteraction', [
+
+                if (warningMessages.length === 0) {
+                  if ($scope.form.schemaForm.$invalid) {
+                    return 'Some of the form controls failed validation';
```
Bit technical. Suggest: "Some of the form entries are invalid." or "Please fix the invalid form entries."

juansaba replied:
Done.

seanlip reviewed (changes requested):
Thanks, @juansaba. I just tried it out, and I think it looks great! I agree with less being more :)

Just a few small code comments, but then this should be good to go.

juansaba committed:
Addressed latest feedback from the PR

juansaba commented:
@seanlip Ok I've addressed the latest comments. Let me know what you think.

As a note, I was hoping to change the region names from "Region{index}" to "Region {index}" i.e. add a space between the word Region and the numbers. I thought it was more human-readable, but I noticed that the validations started failing, cause the space is not allowed in the region name. Is there a valid reason not to allow spaces on region names? I think it would be an improvement in usability to allow them. We don't need to change any of that in this PR though.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -287,6 +299,45 @@ oppia.controller('StateInteraction', [
                 stateCustomizationArgsService.displayed = {};
               };
 
+              $scope.isSaveInteractionEnabled = function() {
```
I'm not sure what enabling a "save interaction" means -- perhaps `areCustomizationArgsSaveable` or `isInteractionSaveable`?

juansaba replied:
Maybe it wasn't obvious but this is talking about the button, i.e. isSaveInteractionButtonEnabled. How do you think about that name? I don't feel too good about "Saveable".

seanlip replied:
isSaveInteractionButtonEnabled sounds good to me.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/state_editor_interaction.html:
```diff
@@ -151,3 +156,12 @@ <h3 ng-if="stateInteractionIdService.displayed && (customizationModalReopened ||
+    background-position: 6px;
+    padding-left: 32px;
+    background-repeat: no-repeat;
+    background-image: url('../assets/common/images/general/warning.png');
```
Just a thought -- did you consider using a material design or fontawesome warning icon instead? We're trying to move away a bit from uploading our own images, since the fonts are already loaded in the page. It also avoids complications with the URL base needing to change in production. (Sorry, I should have noticed this earlier.)

juansaba replied:
Would you mind doing this in a follow up PR?

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/state_editor_interaction.html:
```diff
@@ -151,3 +156,12 @@ <h3 ng-if="stateInteractionIdService.displayed && (customizationModalReopened ||
   </div>
 </script>
+<style>
+  .oppia-interaction-customization-save-disabled {
```
.oppia-save-interaction-button-disabled might be a better name, since you're disabling the "'save interaction' button".

juansaba replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -287,6 +299,45 @@ oppia.controller('StateInteraction', [
+
+              // This function assumes that there are customization args and so
+              // does not do any checks for it.
+              $scope.areCustomizationArgsValid = function() {
```
Why does this need to be a separate function? Could it be combined with the previous one?

juansaba replied:
Yeah, after all the refactoring, this doesn't need to be its own function. It's only used in one place.

seanlip reviewed (changes requested):
Looking good, thanks @juansaba! I noticed a few more small things though.

Re the region names, I agree with you that having a space is more readable. I don't actually know why spaces are currently not allowed, and I think that we can change that in a separate PR. (Might be worth looking up the history in github to see if there was any original context for the restriction, just in case.)

juansaba committed:
Addressing more comments

juansaba commented:
@seanlip Addressed your latest comments. Let me know ;)

juansaba commented:
@seanlip Also I saw failures here:
https://travis-ci.org/oppia/oppia/builds/235095734?utm_source=github_status&utm_medium=notification

But I didn't really find anything obvious pointing to my change. Help?

seanlip commented:
Thanks! Everything looks good; only the warning icon is left (as we discussed).

(Also, the e2e tests now pass, so don't worry about those.)

juansaba committed:
Use material icons for the warning sign

juansaba commented:
@seanlip I've done the change to use material icons for the warning sign. Let me know.

juansaba commented:
@seanlip Oh and just so that you don't have to pull up the app, here's what it looks like now:

![image](https://cloud.githubusercontent.com/assets/4760286/26377200/2bff2e6e-3fc4-11e7-89d7-749edb6111d7.png)

And here's what it used to look like:

![image](https://cloud.githubusercontent.com/assets/4760286/26377244/58cf9b04-3fc4-11e7-89d0-2b82f33ae308.png)

seanlip commented:
Thanks @juansaba! LGTM.
